### PR TITLE
allow populating default value when field is missing

### PR DESCRIPTION
**Pull request by laurenhedberg**
* extend elasticsearch functionality to utilize 'missing' search param - populating a default value when field is missing
* handle searching for missing terms by searching for nil

### CODE REVIEW

File: .rakeTasks [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Settings><!--This file was automatically generated by Ruby plugin.
```
Comment (omeedrj) [inline .rakeTasks]: 🤔 I'm wondering which command added this file. If you delete this file, will it come back after running any rake command?

File: Gemfile [hunk 2]
```diff
@@ -9,6 +9,6 @@ gem "hobo_support",         "2.0.1", git: "git@github.com:Invoca/hobosupport",
 gem "large_text_field",              git: "git@github.com:Invoca/large_text_field.git",  ref: "2efc950395352bf8b7f45891122f6bc42b171526"
 
 group :development do
-  gem 'rubocop', '0.54.0', require: false
+  gem 'rubocop', '0.74.0', require: false
```
Comment (omeedrj) [inline Gemfile:12]: Thank you for addressing this! (along with the ruby version bump)

File: lib/elasticsearch_models/base.rb [hunk 3]
```diff
@@ -2,9 +2,8 @@
 
 module ElasticsearchModels
   class Base < ElasticsearchModels::Aggregate
-    include ElasticsearchModels::DeepSquash
-
     class CreateError < StandardError; end
+    include ElasticsearchModels::DeepSquash
```
Comment (omeedrj) [inline lib/elasticsearch_models/base.rb]: Do you know why was this moved? (I assume) the behavior is the same here as it's original location.

In general it's best to only include changes for refactoring or changing functionality, so that PR's always stay as minimal as necessary.

File: .ruby-version [hunk 4]
```diff
@@ -1 +1 @@
-2.4.2
+2.6.1
```
Comment (omeedrj) [inline .ruby-version:1]: Leaving this comment here, but it mainly relates to other files:
It's a large comment **but** hopefully it doesn't actually take too long to address.

Since you're changing functionality we should make sure to account for some organizational steps in the gem

1. Please bump the gem version in `lib/elasticsearch_models/version.rb`
  a. The summary in https://semver.org/ is generally considered best practice for versioning. Because we're updating the expected ruby version and because `nil` is now being treated different than before, we should do a major version bump to `1.0.0`.
2. Please make sure to update the `README.md` by adding examples and descriptions for the changes in functionality.
  a.  For the `where(field: nil)` functionality. It's important to point out in the comment(s) for the example(s) that searching for a missing field will return documents that have `nil` or empty array `[]` for that key, [as described in the elasticsearch documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-exists-query.html#exists-query-top-level-params).
  b. For the `missing` arg that can be passed in `_aggs`.
3. This currently does not exist in this repo, **but** we (Invoca) should work towards adhering to open source best practices, so please add a `CHANGELOG.md` file that describes the changes being made in `1.0.0` (functional changes and ruby version change, don't necessarily need to include details about updating the rubocop version).
  a. Given that the changelog won't cover all prior versions, you can avoid needing to fill out all version info by describing that the changelog was added in `1.0.0`. The [changelog from the aggregate gem](https://github.com/Invoca/aggregate/blob/50bcd40395985b7d3f2a8f8b6a170f54404ec680/CHANGELOG.md#L1) is a good reference for this, see version `0.2.0`.
  b. Also please add a section to the very top of the `README.md` indicating that when working in this repo it is required to update the version, readme, and changelog when making changes. Here's what I would suggest:
```
# ElasticsearchModels
Model representation for Elasticsearch documents based on Aggregate

## Working In This Repo
When working and making changes in this repo it is required to...
* Update the `README.md` relative to your changes (API examples are always recommended!).
* Bump the gem version appropriately (see MAJOR.MINOR.PATCH summary below).
* Add a corresponding version entry describing what is being Added, Changed, and or Removed in `CHANGELOG.md`.

### MAJOR.MINOR.PATCH Summary (from https://semver.org/)
Given a version number MAJOR.MINOR.PATCH, increment the:
MAJOR version when you make incompatible API changes.
MINOR version when you add functionality in a backwards compatible manner.
PATCH version when you make backwards compatible bug fixes.
```

File: lib/elasticsearch_models/query/aggregation_term.rb [hunk 5]
```diff
@@ -8,12 +8,13 @@ class AggregationTerm
       DEFAULT_ORDER_BY = "_count"
       DEFAULT_ORDER_DIRECTION = "desc"
 
-      def initialize(field:, size: nil, order: nil, partition: nil, num_partitions: nil)
+      def initialize(field:, size: nil, order: nil, partition: nil, num_partitions: nil, missing: nil)
```
Comment (omeedrj) [inline lib/elasticsearch_models/query/aggregation_term.rb:11]: The changes in this class are missing test coverage. Sorry I didn't point this out when you had questions earlier.

In general, any time a class has changes in functionality there should be corresponding changes/additions to its tests.

File: log/test.log [hunk 6]
```diff
@@ -0,0 +1,55 @@
+{"app":{"@timestamp":"2020-01-07T10:54:47.429-08:00","severity":"DEBUG","host":"nblhedberg.local","pid":20017,"tid":"oufqr6x40","service_name":null,"logfile":"/Users/lhedberg/invoca/elasticsearch_models/log/test.log","fiber":null,"message":"Adapter: considering polling ...","timestamp":"2020-01-07T10:54:47.429-08:00","progname":null}}
```
Comment (omeedrj) [inline log/test.log]: Looks like we should add `/log/` to the `.gitignore` so this won't be included

File: spec/query/aggregations_spec.rb [hunk 7]
```diff
@@ -185,6 +185,33 @@
             end
           end
 
+          context "And value is a Hash with missing option" do
+            let(:condition) { { field: "some.field.keyword", size: 10_000, order: "_key", aggs: { field: "some.field.id", missing: 1 } } }
+
+            it "also builds sub aggregation terms with provided missing option" do
```
Comment (omeedrj) [inline spec/query/aggregations_spec.rb:218]: I'd suggest adding test cases for...

* Ensuring that `missing` can be passed for the top level `field` (`"some.field.keyword"`)
  * `{ field: "some.field.keyword", missing: "N/A", size: 10_000, order: "_key", aggs: { field: "some.field.id" } }`
* Ensuring that `missing` can be passed for the top level `field` (`"some.field.keyword"`) and to additional fields
  * `{ field: "some.field.keyword", missing: "N/A", size: 10_000, order: "_key", aggs: { field: "some.field.id", missing: 1 } }`

File: spec/query/aggregations_spec.rb [hunk 8]
```diff
@@ -195,7 +222,7 @@
                     terms: {
                       field: "some.field.keyword",
                       size: 10_000,
-                      order: [{ "_key" => "desc" }]
+                      order: [{ "_key" => "desc" }],
```
Comment (omeedrj) [inline spec/query/aggregations_spec.rb]: nit: The extra `,` isn't needed

File: spec/base_spec.rb [hunk 9]
```diff
@@ -798,6 +798,7 @@ class DummyReplacedModel < DummyElasticSearchModel; end
             @dummy_model2 = DummyElasticSearchModel.create!(my_string: "Hello2")
             @dummy_model3 = DummyElasticSearchModel.create!(my_string: "Hey1", my_other_string: "Hello3")
             @dummy_model4 = DummyElasticSearchModel.create!(my_string: "Hey2", my_other_string: "Hello4")
+            @dummy_model5 = DummyElasticSearchModel.create!(my_string: "Whatsup")
```
Comment (omeedrj) [inline spec/base_spec.rb:811]: Adding the comment here but it pertains to the test file itself:

It looks like we're missing functional test coverage to ensure the `missing` option can be passed to AggregationTerm. We should add a `"with :missing option"` context in `describe ".distinct_values"` and add tests for the three cases
* `missing` for the top level field
* `missing` for the additional fields
* `missing` for the top level field and additional fields

Let me know if you have questions or would like to pair for a bit on this 👍

**Review by omeedrj — CHANGES_REQUESTED** [review]
Great work! Let me know if you have questions on any of the comments

File: lib/elasticsearch_models/base.rb (hunk 3 above)
Comment (laurenhedberg) [reply lib/elasticsearch_models/base.rb]: I just moved this around in my testing/ fixing failed cases. No particular reason, I'll move it back

File: .rakeTasks (hunk 1 above)
Comment (laurenhedberg) [reply .rakeTasks]: Hmm... not sure why this was added. I removed and ran rake command - it didn't pop up again. I've deleted now.

File: log/test.log (hunk 6 above)
Comment (omeedrj) [reply log/test.log]: Was this addressed?

File: spec/query/aggregations_spec.rb [hunk 10]
```diff
@@ -185,6 +185,95 @@
...
+
+            it "builds aggregation terms with missing field at the top level" do
+              expected_terms = {
+                aggs: {
+                  "some.field.keyword" => {
+                    terms: {
+                      field: "some.field.keyword",
+                        size: 10000,
+                        order: [{ "_key" => "desc" }],
+                        missing: "N/A"
```
Comment (omeedrj) [inline spec/query/aggregations_spec.rb]: nit: A number of hashes throughout these tests have misaligned keys

File: README.md [hunk 11]
```diff
@@ -171,10 +184,16 @@ String searching using the `_q` parameter is fuzzy by default (`*` on both ends
...
 e.g. `full text` search string will be converted to the following query string `(*full AND text*)`
+
+Querying for nil fields will return documents that have `nil` or an empty array`[]` for the given key
+
 ```ruby
 # Query by full text string searching
 DummyElasticSearchModel.where(_q: "Hello")
 
+# Query by nil fields
+DummyElasticSearchModel.where(_q: { my_string: nil })
```
Comment (omeedrj) [inline README.md]: Important note: the `nil` searching will be applied when passed like so
```
DummyElasticSearchModel.where(my_string: nil)
```

The above behaves like a filter, where as using the `_q` option behaves like fuzzy search.

File: README.md [hunk 12]
```diff
@@ -297,6 +317,9 @@ DummyElasticSearchModel.where(_aggs: { field: "my_string.keyword", order: { "_ke
 # Aggregate on a single field with multiple ordering options
 DummyElasticSearchModel.where(_aggs: { field: "my_string.keyword", order: ["_key", { "_count" => "asc" }] })
 
+# Aggregate on a single field while populating missing fields with the provided default value
+DummyElasticSearchModel.where(_aggs: { field: "my_string.keyword", missing: "N/A" })
```
Comment (omeedrj) [inline README.md:319]: Would you also mind adding an example in the `.distinct_values` section?

```
DummyElasticSearchModel.distinct_values("my_int", missing: 0, additional_fields: [{ field: "my_string", missing: "N/A" }])
```

File: CHANGELOG.md [hunk 13]
```diff
@@ -0,0 +1,5 @@
+# Changelog
+All notable changes to this project will be documented in this file
+## [1.0.0] - 2020-01-13
+### Added
```
Comment (omeedrj) [inline CHANGELOG.md:4]: For version `1.0.0`, please also add 
```
### Added
- Indicate a default `missing` value for aggregations when documents are missing the expected aggregation term.
- Querying for records can now filter for fields that have missing values (`nil` or `[]`)

### Changed
- Updated ruby version from `2.4.2` to `2.6.1`
```

File: lib/elasticsearch_models/base.rb [hunk 14]
```diff
@@ -67,10 +67,14 @@ def count(**params)
 
       def distinct_values(field, additional_fields: [], where: {}, **params)
         field.presence.is_a?(String) or raise ArgumentError, "field must be a present String"
-        additional_fields.all? { |f| f.presence.is_a?(String) } or raise ArgumentError, "additional_fields must all be present Strings"
+        additional_fields.is_a?(Array) or raise ArgumentError, "additional_fields must be an Array"
+        additional_fields.all? do |f|
+          f.presence.is_a?(String) || (f.is_a?(Hash) && f.key?(:field))
+        end or raise ArgumentError, "additional_fields must all be present Strings"
```
Comment (omeedrj) [inline lib/elasticsearch_models/base.rb]: The error message for this exception isn't accurate, maybe `"additional_fields must all be Strings or Hashes with a :field key"`?

File: spec/base_spec.rb [hunk 15]
```diff
@@ -291,8 +291,6 @@ class DummyReplacedModel < DummyElasticSearchModel; end
       end
 
       it "validates the model before attempting to insert to Elasticsearch" do
-        expect { DummyElasticSearchModel.create! }.to raise_error(ActiveRecord::RecordInvalid, "Validation failed: My string must be set")
```
Comment (omeedrj) [inline spec/base_spec.rb:294]: Removing this expectation means the test is no longer testing anything and we're losing test coverage for the `required: true` functionality of class fields.

Is there a way we could update the class without removing the `required: true` and without changing this test?

File: spec/base_spec.rb [hunk 16]
```diff
@@ -1275,19 +1291,87 @@ class DummyReplacedModel < DummyElasticSearchModel; end
...
+                },
+                "This is a test" => {
+                  "my_other_string.keyword" => ["Missing additional field"]
+                },
+                "Hey" => {
+                  "my_other_string.keyword" => ["Missing additional field"]
+                },
+                "Hello again" => {
+                  "my_other_string.keyword" => ["Missing additional field"]
+                }
```
Comment (omeedrj) [inline spec/base_spec.rb]: nit: The hash keys are misaligned

File: spec/base_spec.rb [hunk 17]
```diff
@@ -1275,19 +1291,87 @@ class DummyReplacedModel < DummyElasticSearchModel; end
...
+      context "with top-level field nil" do
+        let(:missing) { "Missing field" }
+
+        it "populates empty top-level field with missing value" do
+          expected_values = { "my_string.keyword" => ["Hello", "This is a test", "Hello again", "Hey", "How are you?", "Missing field"] }
+          expect(distinct_values).to eq(expected_values)
+        end
+      end
+
+      context "with additional field nil", lauren: true do
```
Comment (omeedrj) [inline spec/base_spec.rb]: nit: `lauren: true`

**Review by omeedrj — CHANGES_REQUESTED** [review]
Nice work, a few more comments and then it should be good to go!

File: spec/base_spec.rb (hunk 15 above)
Comment (laurenhedberg) [reply spec/base_spec.rb:294]: So yes, I'm sure there is a way we can do this, but I haven't found a straightforward way to do so that doesn't involve updating a lot of test cases. I've already spent a lot of time on this so if we could go over it together that'd be great.

File: log/test.log (hunk 6 above)
Comment (laurenhedberg) [reply log/test.log]: Oops, sorry I must have missed this earlier. It's been addressed now.

File: spec/base_spec.rb (hunk 15 above)
Comment (omeedrj) [reply spec/base_spec.rb:294]: Looks like you really quickly found a way to make this work, nice 👍

**Review by omeedrj — APPROVED** [review]
💯
